hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -364,6 +372,57 @@ protected List<DatanodeDetails> chooseDatanodesInternal(
     return result;
   }
 
+  /**
+   * Given a list of excluded nodes, check if the rack for each excluded node is
+   * empty after removing the excluded nodes. If it is empty, then the rack
+   * contains only excluded nodes, and we return a list of these racks.
+   * @param excludedNodes List of excluded nodes
+   * @param usedRacksCntMap Map of used racks and their used node count
+   * @return List of racks that contain only excluded nodes or an empty list
+   */
+  private List<Node> findRacksWithOnlyExcludedNodes(
+      List<DatanodeDetails> excludedNodes, Map<Node, Integer> usedRacksCntMap) {
+    if (excludedNodes == null || excludedNodes.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    List<Node> unavailableRacks = new ArrayList<>();
+    Set<Node> excludedNodeRacks = new HashSet<>();
+    for (Node node : excludedNodes) {
+      Node rack = networkTopology.getAncestor(node, RACK_LEVEL);
+      if (rack != null && !usedRacksCntMap.containsKey(rack)) {
+        // Rack for an excluded node and no used nodes on this rack.
+        excludedNodeRacks.add(rack);
+      }
+    }
+    Set<Node> exc = new HashSet<>(excludedNodes);
+    for (Node rack : excludedNodeRacks) {
+      // Dead nodes are removed from the topology, so the node may have a null
+      // rack, or the topology may not know about it, depending on the timing
+      // of when it got removed / excluded.
+      if (rack == null) {
+        continue;
+      }
```

Comment:
Is this check needed given that we're checking for null at lines 393 and 411?

Reply:
Yes, you are correct. The Rack must be non null in the map, as we check it before inserting it. There is no need to check again here.